subsys/usb/class/bluetooth.c

```diff
@@ -0,0 +1,236 @@
+/*
+ * Wireless / Bluetooth USB class
+ *
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+
+#include <usb/usb_device.h>
+#include <usb/usb_common.h>
+#include <usb_descriptor.h>
+
+#include <net/buf.h>
+
+#include <bluetooth/buf.h>
+#include <bluetooth/hci_raw.h>
+#include <bluetooth/l2cap.h>
+
+#define SYS_LOG_LEVEL SYS_LOG_LEVEL_WARNING
+#define SYS_LOG_DOMAIN "usb/bluetooth"
+#include <logging/sys_log.h>
+
+#if !defined(CONFIG_USB_COMPOSITE_DEVICE)
+static u8_t interface_data[64];
+#endif
+
+static K_FIFO_DEFINE(rx_queue);
+
+/* HCI command buffers */
+#define CMD_BUF_SIZE BT_BUF_RX_SIZE
+NET_BUF_POOL_DEFINE(tx_pool, CONFIG_BT_HCI_CMD_COUNT, CMD_BUF_SIZE,
+		    sizeof(u8_t), NULL);
+
+#define BT_L2CAP_MTU 64
+/* Data size needed for ACL buffers */
+#define BT_BUF_ACL_SIZE BT_L2CAP_BUF_SIZE(BT_L2CAP_MTU)
+NET_BUF_POOL_DEFINE(acl_tx_pool, 2, BT_BUF_ACL_SIZE, sizeof(u8_t), NULL);
+
+/* Stack for thread */
+static K_THREAD_STACK_DEFINE(bluetooth_thread_stack, 512);
+static struct k_thread bluetooth_thread_data;
+
+static void bluetooth_status_cb(enum usb_dc_status_code status, u8_t *param)
+{
+	/* Check the USB status and do needed action if required */
+	switch (status) {
+	case USB_DC_ERROR:
+		SYS_LOG_DBG("USB device error");
+		break;
+	case USB_DC_RESET:
+		SYS_LOG_DBG("USB device reset detected");
+		break;
+	case USB_DC_CONNECTED:
+		SYS_LOG_DBG("USB device connected");
+		break;
+	case USB_DC_CONFIGURED:
+		SYS_LOG_DBG("USB device configured");
+		break;
+	case USB_DC_DISCONNECTED:
+		SYS_LOG_DBG("USB device disconnected");
+		break;
+	case USB_DC_SUSPEND:
+		SYS_LOG_DBG("USB device suspended");
+		break;
+	case USB_DC_RESUME:
+		SYS_LOG_DBG("USB device resumed");
+		break;
+	case USB_DC_UNKNOWN:
+	default:
+		SYS_LOG_DBG("USB unknown state");
+		break;
+	}
+}
+
+static void bluetooth_thread(void)
+{
+	SYS_LOG_DBG("Start USB Bluetooth thread");
+
+	while (true) {
+		struct net_buf *buf;
+
+		buf = net_buf_get(&rx_queue, K_FOREVER);
+
+		switch (bt_buf_get_type(buf)) {
+		case BT_BUF_EVT:
+			usb_transfer_sync(CONFIG_BLUETOOTH_INT_EP_ADDR,
+					  buf->data, buf->len,
+					  USB_TRANS_WRITE | USB_TRANS_NO_ZLP);
+			break;
+		case BT_BUF_ACL_IN:
+			usb_transfer_sync(CONFIG_BLUETOOTH_IN_EP_ADDR,
+					  buf->data, buf->len,
+					  USB_TRANS_WRITE | USB_TRANS_NO_ZLP);
+			break;
+		default:
+			SYS_LOG_ERR("Unknown type %u", bt_buf_get_type(buf));
+			break;
+		}
+
+		net_buf_unref(buf);
+	}
+}
+
+static void bluetooth_bulk_out_callback(u8_t ep,
+					enum usb_dc_ep_cb_status_code ep_status)
+{
+	struct net_buf *buf;
+	u32_t len;
+
+	SYS_LOG_DBG("ep %x status %d", ep, ep_status);
+
+	/* Read number of bytes to read */
+	usb_read(ep, NULL, 0, &len);
+
```

Comment:
We could use transfer API for reading a full HCI packet as well. Actually here we don't read a full HCI packet but a chunk of hci packet which is then transmitted to the Bluetooth driver via bt_send. In case of h4/UART (USB->Zephyr->UART->BTController), this works since the packet is retransmitted via UART and then re-aggregated on controller side, but I would expect that bt_send provides a complete HCI packet for larger compatibility.

Reply:
I think I will try to use transfer for read also.

Reply:
Now I think that usb_read() API is clear and much easy to understand then usb_transfer(READ), I leave it for now as is.

Reply:
If it works no problem, we will potentially change it later.